AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt

```diff
@@ -0,0 +1,185 @@
+
+import android.app.PendingIntent
+import android.appwidget.AppWidgetManager
+import android.appwidget.AppWidgetProvider
+import android.content.Context
+import android.content.Intent
+import android.widget.RemoteViews
+import android.widget.RemoteViewsService
+import android.widget.Toast
+import androidx.annotation.LayoutRes
+import androidx.annotation.RequiresApi
+import androidx.core.os.BuildCompat
+import com.example.android.appwidget.ItemsCollectionRemoteViewsFactory.Companion.EXTRA_VIEW_ID
+import com.example.android.appwidget.ItemsCollectionRemoteViewsFactory.Companion.REQUEST_CODE
+import com.example.android.appwidget.ItemsCollectionRemoteViewsFactory.Companion.REQUEST_CODE_FROM_COLLECTION_WIDGET
+
+/**
+ * Implementation of App Widget functionality that demonstrates the difference of how the list of
+ * items are inflated with API level 31 and older API levels.
+ * This widget also demonstrates the compound buttons (Checkbox, RadioButton, Switch), that are now
+ * supported in widgets starting from API level 31.
+ */
+class ItemsCollectionAppWidget : AppWidgetProvider() {
+
+    override fun onUpdate(
+        context: Context,
+        appWidgetManager: AppWidgetManager,
+        appWidgetIds: IntArray
+    ) {
+        val remoteViews = RemoteViews(context.packageName, R.layout.widget_items_collection)
+        if (BuildCompat.isAtLeastS()) {
+            val collectionItems =
+                ItemsCollectionRemoteViewsFactory.getRemoteCollectionItems(context)
+            remoteViews.setRemoteAdapter(R.id.items_list_view, collectionItems)
+        } else {
+            remoteViews.setRemoteAdapter(
+                R.id.items_list_view,
+                Intent(context, ItemsCollectionRemoteViewsService::class.java)
+            )
+        }
+        appWidgetManager.updateAppWidget(appWidgetIds, remoteViews)
+    }
+
+    override fun onReceive(context: Context?, intent: Intent?) {
+        super.onReceive(context, intent)
+        if (BuildCompat.isAtLeastS() &&
+            intent?.extras?.getInt(REQUEST_CODE) == REQUEST_CODE_FROM_COLLECTION_WIDGET
+        ) {
+            intent.extras?.let {
+                // The checked state is stored in the Bundle in the Intent
+                val checked = it.getBoolean(
+                    RemoteViews.EXTRA_CHECKED,
+                    false
+                )
+                Toast.makeText(
+                    context,
+                    "ViewId : ${it.getInt(EXTRA_VIEW_ID)}'s checked status is now : $checked",
+                    Toast.LENGTH_SHORT
+                ).show()
+            }
+        }
+    }
+}
+
+class ItemsCollectionRemoteViewsService : RemoteViewsService() {
+
+    override fun onGetViewFactory(data: Intent?): RemoteViewsFactory {
+        return ItemsCollectionRemoteViewsFactory(applicationContext)
+    }
+}
+
+class ItemsCollectionRemoteViewsFactory(private val context: Context) :
+    RemoteViewsService.RemoteViewsFactory {
+
+    companion object {
```

Comment:
According to the Kotlin style guide, the companion object should be at the end of the class defintion: https://kotlinlang.org/docs/coding-conventions.html#class-layout

Reply:
Done.